File: 9.py
#!/usr/bin/python3

import re

def recurse(sequence):
    if set(sequence) == set([0]):
        return [0, 0]
    new_sequence = [x - y for x, y in zip(sequence[1:], sequence[:-1])]
    val = recurse(new_sequence)
    return [sequence[0] - val[0], sequence[-1] + val[1]]

with open('inputs/9') as f:
    data = f.read().rstrip('\n').split('\n')
sums = [0,0]
for line in data:
    n = list(map(int, re.findall(r'[0-9\-]+', line)))
    sums = [x + y for x, y in zip(sums, recurse(n))]
print(f"Part 1: {sums[1]}\nPart 2: {sums[0]}")

File: 9.rb
def recurse(arr)
  if arr.uniq.size == 1 && arr[0].zero?
    return [0,0]
  end
  result = recurse(arr[..-2].zip(arr[1..]).map{|a| a=a[1]-a[0]})
  return [arr[0]-result[0], arr[-1] + result[1]]
end

lines = IO.readlines("inputs/9").map(&:strip)
ans = [0,0]
lines.each { |line|
  fields = line.split(/ /).map(&:to_i)
  ans = [ans, recurse(fields)].transpose.map(&:sum)
}
puts "Part 1: #{ans[1]}"
puts "Part 2: #{ans[0]}"
